### PR TITLE
Add ephemeral AbstractNode support for graphastar.Path

### PR DESCRIPTION
**Pull request by minkezhang**
Ephemeral AbstractNode instances are created as part of the SG-generation step (Botea 2004). These nodes are connected to every other (non-ephemeral) AbstractNode in the same cluster, are necessary for finding the optimal abstract path.

### CODE REVIEW

**Review by RyanzpLee — APPROVED** [review]

File: lib/hpf/graph.go [hunk 1]
```diff
@@ -95,70 +96,143 @@ func BuildGraph(tm *tile.Map, tileDimension *rtsspb.Coordinate) (*Graph, error)
...
-						return nil, err
-					}
-
-					if e != nil {
-						g.EdgeMap.Add(e)
-					}
-				}
+		for _, n := range nodes {
+			if err := connect(tm, g, utils.MC(n.GetTileCoordinate())); err != nil {
+				return nil, err
```
[the remark this answers is not in the document]
Comment (minkezhang) [reply lib/hpf/graph.go:101]: https://golang.org/ref/spec#Short_variable_declarations

This is shorthand for `var x xType = xType{}`; that is, it's basically C++'s `auto` keyword.

File: lib/hpf/graph.go [hunk 2]
```diff
@@ -95,70 +96,143 @@ func BuildGraph(tm *tile.Map, tileDimension *rtsspb.Coordinate) (*Graph, error)
...
 	}
 
 	return g, nil
 }
 
-// buildIntraEdge constructs a single AbstractEdge instance with the correct
-// traversal cost between two underlying AbstractNode objects. The cost
-// function is calculated from the tile.Map entity, which holds information
-// on e.g. the terrain information of the map.
-func buildIntraEdge(tm *tile.Map, cm *cluster.Map, n1, n2 *rtsspb.AbstractNode) (*rtsspb.AbstractEdge, error) {
```
[the remark this answers is not in the document]
Comment (minkezhang) [reply lib/hpf/graph.go:121]: `EdgeType.EDGE_TYPE_INTRA` is an edge that connects different `AbstractNode` instances within the same cluster. We need to know how to get through clusters.

The `EdgeType` is just a property of `AbstractEdge`. It's defined in the protobuf definition [lib/proto/structs.proto](https://github.com/minkezhang/rts-pathing/blob/4c93a4effc170922474c2b8e7ceceb69e5fb690a/lib/proto/structs.proto#L62).

File: lib/hpf/graph.go [hunk 3]
```diff
@@ -95,70 +96,143 @@ func BuildGraph(tm *tile.Map, tileDimension *rtsspb.Coordinate) (*Graph, error)
...
-	c1, err := cluster.ClusterCoordinateFromTileCoordinate(cm, utils.MC(n1.GetTileCoordinate()))
+// connect takes as input an AbstractNode, builds all possible INTRA_EDGE
+// AbstractEdge instances within the same cluster, and inserts them into
+// the Graph.
+//
+// connect does not rebuild edges if they already exist between two nodes, and
+// does not create an edge between two ephemeral AbstractNode instances.
+func connect(tm *tile.Map, g *Graph, t utils.MapCoordinate) error {
```
[the remark this answers is not in the document]
Comment (minkezhang) [reply lib/hpf/graph.go:115]: The input of this function gives the coordinate of an existing `AbstractNode`. The function computes and inserts all edges from `AbstractNode` instances within the same cluster as the input, to the input node.

File: lib/hpf/graph_test.go [hunk 4]
```diff
@@ -552,3 +428,122 @@ func TestGraphGetNeighbors(t *testing.T) {
...
+			t.Fatalf("Add() = %v, want = nil", err)
+		}
+	}
+
+	testConfigs := []struct {
+		name string
+		tm   *tile.Map
+		g    *Graph
+		t    *rtsspb.Coordinate
+		want []*rtsspb.AbstractEdge
```
[the remark this answers is not in the document]
Comment (minkezhang) [reply lib/hpf/graph_test.go:490]: Yes. Golang prefers [table-driven tests](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests), and the variable `want` is typically reserved for the expected output.

File: lib/hpf/graph.go [hunk 5]
```diff
@@ -95,70 +96,143 @@ func BuildGraph(tm *tile.Map, tileDimension *rtsspb.Coordinate) (*Graph, error)
...
+				Weight:      cost,
+			})
+		}
 	}
-	tileDimension, err := cluster.TileDimension(cm, c1)
+
+	return nil
+}
+
+// AddEphemeralNode adds a temporary AbstractNode to the Graph and connects it
```
[the remark this answers is not in the document]
Comment (minkezhang) [reply lib/hpf/graph.go:177]: An ephemeral node is subclass of AbstractNode -- we are creating these as part of SG-generation. These nodes represent the individual unit paths -- it doesn't make sense to have multiple unit paths be dependent on one another. These nodes will only live for as long as we need to perform the pathing calculation, and therefore are different from the other static nodes (which may still undergo changes due to map changes, but are always expected to be part of the underlying pathing).

**Pull request merged**